internal/exec.go

```diff
@@ -0,0 +1,79 @@
+package internal
+
+import (
+	"bytes"
+	"context"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/cloudogu/cesapp-lib/core"
+	corev1 "k8s.io/api/core/v1"
+
+	k8sv1 "github.com/cloudogu/k8s-dogu-operator/api/v1"
+)
+
+// ExecPod provides methods for instantiating and removing an intermediate pod based on a Dogu container image.
+type ExecPod interface {
+	// Create adds a new exec pod to the cluster.
+	Create(ctx context.Context) error
+	// Delete deletes the exec pod from the cluster.
+	Delete(ctx context.Context) error
+	// PodName returns the name of the pod.
+	PodName() string
+	// ObjectKey returns the ExecPod's K8s object key.
+	ObjectKey() *client.ObjectKey
+	// Exec runs the provided command in this execPod
+	Exec(ctx context.Context, cmd ShellCommand) (out string, err error)
+}
+
+// ExecPodFactory provides functionality to create ExecPods.
+type ExecPodFactory interface {
+	// NewExecPod creates a new ExecPod.
+	NewExecPod(execPodFactoryMode ExecPodVolumeMode, doguResource *k8sv1.Dogu, dogu *core.Dogu) (ExecPod, error)
+}
+
+// ExecPodVolumeMode indicates whether to mount a dogu's PVC (which only makes sense when the dogu was already
+// installed).
+type ExecPodVolumeMode int
+
+const (
+	// VolumeModeInstall indicates to not mount a dogu's PVC.
+	VolumeModeInstall ExecPodVolumeMode = iota
+	// VolumeModeUpgrade indicates to mount a dogu's PVC.
+	VolumeModeUpgrade
+)
+
+// CommandExecutor is used to execute commands in pods and dogus
+type CommandExecutor interface {
+	// ExecCommandForDogu executes a command in a dogu.
+	ExecCommandForDogu(ctx context.Context, resource *k8sv1.Dogu, command ShellCommand, expectedStatus PodStatus) (*bytes.Buffer, error)
+	// ExecCommandForPod executes a command in a pod that must not necessarily be a dogu.
+	ExecCommandForPod(ctx context.Context, pod *corev1.Pod, command ShellCommand, expectedStatus PodStatus) (*bytes.Buffer, error)
+}
+
+// ShellCommand represents a command that can be executed in the shell of a container.
+type ShellCommand interface {
+	// CommandWithArgs returns the commands and its arguments in a way suitable for execution.
+	CommandWithArgs() []string
+}
+
+// PodStatus describes a state in the lifecycle of a pod.
+type PodStatus string
+
+const (
+	// ContainersStarted means that all containers of a pod were started.
+	ContainersStarted PodStatus = "started"
+	// PodReady means that the readiness probe of the pod has succeeded.
+	PodReady PodStatus = "ready"
+)
+
```

Comment:
Collecting interfaces in an internal package is one thing, I fully understand the reason. But here we have actual values which from outside are referenced as `internal.PodReady`. The package helps to identify the business value of a thing. And now, by moving here, the business context is gone.

If we are not careful we might encourage other developers to add further values here, also void of business context. So how do we solve this situation?